compactor: configure via cluster settings

tbg opened the pull request:
The compactor can cause instability or degrade performance, and it does
so in an intransparent way, partially because its compactions are
low-priority and are thus regularly pending for upwards of ten minutes,
without being able to tell at which point work is being done. It's
important to be able to change its behavior, and in particular to
disable it, whether that's because of a known problem or a suspected
one. This wasn't possible at all previously. Now it can be achieved by
setting `compactor.threshold_bytes` and `max_record_age` to a very high
and low value, respectively.

Release note (general change): Introduced cluster settings to configure
the compactor.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25397)
<!-- Reviewable:end -->

bdarnell commented:
> Now it can be achieved by
> setting `compactor.threshold_bytes` and `max_record_age` to a very high
> and low value, respectively.

To really disable the compactor, you'd also have to set bytes_threshold and bytes_available_threshold too. Should this just be a boolean setting to disable the compactor completely? It seems like the issues we've seen with the compactor won't generally be fixed by tweaking the numbers to make it run a little less often, and if the main anticipated use case is to disable it completely it's better to add a setting that just does that than to expose all the internal complexity of the compactor. 

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/compactor/settings.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFIgDf7VdZQ2LAwwcT:-LCFIgDf7VdZQ2LAwwcU:biypd9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L25)):*
> ```Go
> )
> 
> const (
> ```

Can you inline these into the setting declarations?

---

*[pkg/storage/compactor/settings.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFGINn6lTBsOKeCuHQ:-LCFGINn6lTBsOKeCuHR:b-hw6nc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L33)):*
> ```Go
> )
> 
> func validatePercentageFloat(v float64) error {
> ```

s/PercentageFloat/Fraction/? Percentages are between 0 and 100. 

---

*[pkg/storage/compactor/settings.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFGV3CBMJZq4iVU8bJ:-LCFGV3CBMJZq4iVU8bK:b-baiuka) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L35)):*
> ```Go
> func validatePercentageFloat(v float64) error {
> 	err := errors.New("value not between zero and one")
> 	if math.IsNaN(v) || math.IsInf(v, 1) || math.IsInf(v, -1) || v < 0 || v > 1 {
> ```

I think if you inverted this and did `if v >= 0 && v <= 1 { return nil }` you could avoid the nan/inf checks.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-:-LCFJIXsAKALUsby0aXw:boj0fd5)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I introduced a `compactor.enabled` setting (with associated test) and hid the other ones. I agree that we hopefully don't have to fine-tune those too much, but if there's any hope of making the compactor something that can be turned on in general, we'll have to figure out sensible thresholds, and that will be easier if there are cluster settings for it.

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/compactor/settings.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFIgDf7VdZQ2LAwwcT:-LCFdVL34NBppPcE75cN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L25)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can you inline these into the setting declarations?
</blockquote></details>

Done.

---

*[pkg/storage/compactor/settings.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFGINn6lTBsOKeCuHQ:-LCFdVpk5HwrNPQ4Nc6n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L33)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/PercentageFloat/Fraction/? Percentages are between 0 and 100. 
</blockquote></details>

Done.

---

*[pkg/storage/compactor/settings.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCFGV3CBMJZq4iVU8bJ:-LCFdW9y5eAnyNVuQXG9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/594a778c79f9f3e078cef6e0be57033bd06ecf1a/pkg/storage/compactor/settings.go#L35)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think if you inverted this and did `if v >= 0 && v <= 1 { return nil }` you could avoid the nan/inf checks.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-:-LCFcm_Z2uYGUgEiVz1K:b-esca2i)*
<!-- Sent from Reviewable.io -->

a-robinson reviewed (commented):
> its compactions are
> low-priority and are thus regularly pending for upwards of ten minutes,
> without being able to tell at which point work is being done.

Did you confirm this?

tschottdorf commented:
> Did you confirm this?

What exactly? I had forgotten about the `rocksdb.compactions` metric when I looked at these 13+ minute compactions. It's possible that that metric would've helped pinpoint the compactions more specifically, but I can't access that cluster any more.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---

*[pkg/storage/compactor/settings.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCQfHgmDRkkDFcNcYZo:-LCQfHgmDRkkDFcNcYZp:bl6lzl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f551036fa1347aad3bdc9a3dfc3fc11498cae02/pkg/storage/compactor/settings.go#L43)):*
> ```Go
> var enabled = settings.RegisterBoolSetting(
> 	"compactor.enabled",
> 	"when false, the compactor will immediately discard all past and future suggestions",
> ```

"immediately discards suggestions" is an implementation detail. Since this is a public-facing setting, say something like "when false, the system will reclaim space occupied by deleted data less aggressively". 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-:-LCQfZ8u9s0y3vzfq5jn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
> > > its compactions are
> > > low-priority and are thus regularly pending for upwards of ten minutes,
> > > without being able to tell at which point work is being done.
> >
> > Did you confirm this?
>
> What exactly? I had forgotten about the rocksdb.compactions metric when I looked at these 13+ minute compactions. It's possible that that metric would've helped pinpoint the compactions more specifically, but I can't access that cluster any more (at this very moment).

I mean did you confirm that what's happening is that the compactions are pending for most of their "duration"? Rather than that the compactions are running in the background the whole time but just really slowly?

tschottdorf commented:
> I mean did you confirm that what's happening is that the compactions are pending for most of their "duration"? Rather than that the compactions are running in the background the whole time but just really slowly?

No, so far I only observed that the compaction queue's calls to RocksDB take >10 minutes sometimes. I assume that they are internally blocked by RocksDB and apply in a short burst, but I don't have proof for that. The RocksDB logging change together with logspy can hopefully let us peek at this from up close once we see it in action.

TFTR!

bors r=bdarnell

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/compactor/settings.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-LCQfHgmDRkkDFcNcYZo:-LCReif45n2m1152mvH2:byccrv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f551036fa1347aad3bdc9a3dfc3fc11498cae02/pkg/storage/compactor/settings.go#L43)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"immediately discards suggestions" is an implementation detail. Since this is a public-facing setting, say something like "when false, the system will reclaim space occupied by deleted data less aggressively". 
</blockquote></details>

I'm not sure I'm generally on board with giving these internal knobs vague descriptions (that potentially invite users to fiddle with them), but I made the change you suggested.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25397#-:-LCRf7PG5conAOfTOFea:b-vwhzk5)*
<!-- Sent from Reviewable.io -->

tbg committed:
compactor: configure via cluster settings

The compactor can cause instability or degrade performance, and it does
so in an intransparent way, partially because its compactions are
low-priority and are thus regularly pending for upwards of ten minutes,
without being able to tell at which point work is being done. It's
important to be able to change its behavior, and in particular to
disable it, whether that's because of a known problem or a suspected
one. This wasn't possible at all previously. Now it can be achieved by
setting `compactor.enabled` to `false`; the advanced settings have
been marked as hidden.

Release note (general change): Introduced cluster settings to configure
the compactor.

tschottdorf commented:
Hadn't regenerated settings.html.

bors r=bdarnell

jseldess commented:
@tschottdorf, @dt, why weren't these new settings automatically added to the generated settings doc? https://github.com/cockroachdb/cockroach/blob/master/docs/generated/settings/settings.html

tschottdorf commented:
@jseldess they were marked as hidden.

On Fri, Jul 6, 2018 at 9:40 PM Jesse Seldess <notifications@github.com>
wrote:

> @tschottdorf <https://github.com/tschottdorf>, @dt <https://github.com/dt>,
> why weren't these new settings automatically added to the generated
> settings doc?
> https://github.com/cockroachdb/cockroach/blob/master/docs/generated/settings/settings.html
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/25397#issuecomment-403153540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135OWkUdHoBJbHhEG8DFBTdSnrEn1bks5uD9lBgaJpZM4T5ZwI>
> .
>
-- 

-- Tobias

jseldess commented:
I see, thanks. I guess I was expecting them to be documented since they were announced via release notes.

dt commented:
I'm generally 👎 on hidden settings -- settings in general should just be considered "you should know what you're doing before changing this" so hiding should be reserved for only the most extreme danger settings where we think no one should ever set it. / 2¢.

tschottdorf commented:
I'm fine with that, but once we remove a setting it'll haunt people who
used it with annoying log messages.

On Tue, Jul 10, 2018, 11:41 PM David Taylor <notifications@github.com>
wrote:

> I'm generally 👎 on hidden settings -- settings in general should just be
> considered "you should know what you're doing before changing this" so
> hiding should be reserved for only the most extreme danger settings where
> we think no one should ever set it. / 2¢.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/25397#issuecomment-403975137>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135MB5K5PlTqMVyqYWcEbNCNla7x46ks5uFR9xgaJpZM4T5ZwI>
> .
>
-- 

-- Tobias

jseldess commented:
The docs prefaces the settings with this warning, which I think is strong enough?

> Many cluster settings are intended for tuning CockroachDB internals. Before changing these settings, we strongly encourage you to discuss your goals with Cockroach Labs; otherwise, you use them at your own risk.

dt commented:
NB: when we remove a setting we are supposed to put its name in the `retired` slice, which a) prevents accidental re-use of the name and b) suppresses log-spam about that setting.
